util/pingpong.c

```diff
@@ -0,0 +1,1940 @@
+}
+
+static inline void ft_start(struct ct_pingpong *ct)
+{
+	ct->opts.options |= FT_OPT_ACTIVE;
+	clock_gettime(CLOCK_MONOTONIC, &(ct->start));
+}
+
+static inline void ft_stop(struct ct_pingpong *ct)
+{
+	clock_gettime(CLOCK_MONOTONIC, &(ct->end));
+	ct->opts.options &= ~FT_OPT_ACTIVE;
+}
+
+static int ft_check_opts(struct ct_pingpong *ct, uint64_t flags)
+{
+	return (ct->opts.options & flags) == flags;
+}
+
+void ft_fill_buf(void *buf, int size)
+{
+	char *msg_buf;
+	int msg_index;
+	static unsigned int iter = 0;
+	int i;
+
+	msg_index = ((iter++)*INTEG_SEED) % integ_alphabet_length;
+	msg_buf = (char *)buf;
+	for (i = 0; i < size; i++) {
+		msg_buf[i] = integ_alphabet[msg_index++];
+		if (msg_index >= integ_alphabet_length)
+			msg_index = 0;
+	}
+}
+
+int ft_check_buf(void *buf, int size)
+{
+	char *recv_data;
+	char c;
+	static unsigned int iter = 0;
+	int msg_index;
+	int i;
+
+	msg_index = ((iter++)*INTEG_SEED) % integ_alphabet_length;
+	recv_data = (char *)buf;
+
+	for (i = 0; i < size; i++) {
+		c = integ_alphabet[msg_index++];
+		if (msg_index >= integ_alphabet_length)
+			msg_index = 0;
+		if (c != recv_data[i])
+			break;
+	}
+	if (i != size) {
+		printf("Error at iteration=%d size=%d byte=%d\n",
+			iter, size, i);
+		return 1;
+	}
+
+	return 0;
+}
+
+uint64_t ft_init_cq_data(struct fi_info *info)
+{
+	if (info->domain_attr->cq_data_size >= sizeof(uint64_t)) {
+		return 0x0123456789abcdefULL;
+	} else {
+		return 0x0123456789abcdef &
+			((0x1ULL << (info->domain_attr->cq_data_size * 8)) - 1);
+	}
+}
+
+static void ft_cq_set_wait_attr(struct ct_pingpong *ct)
+{
+	ct->cq_attr.wait_obj = FI_WAIT_NONE; // NEEDED ? SNIP ??
+}
+
+static void ft_cntr_set_wait_attr(struct ct_pingpong *ct)
+{
+	ct->cntr_attr.wait_obj = FI_WAIT_NONE; // NEEDED ? SNIPP ??
+}
+
+static uint64_t ft_caps_to_mr_access(uint64_t caps)
+{
+	uint64_t mr_access = 0;
+
+	if (caps & (FI_MSG | FI_TAGGED)) {
+		if (caps & FT_MSG_MR_ACCESS)
+			mr_access |= caps & FT_MSG_MR_ACCESS;
+		else
+			mr_access |= FT_MSG_MR_ACCESS;
+	}
+
+	if (caps & (FI_RMA | FI_ATOMIC)) {
+		if (caps & FT_RMA_MR_ACCESS)
+			mr_access |= caps & FT_RMA_MR_ACCESS;
+		else
+			mr_access |= FT_RMA_MR_ACCESS;
+	}
+
+	return mr_access;
+}
+
+/*
+ * Include FI_MSG_PREFIX space in the allocated buffer, and ensure that the
+ * buffer is large enough for a control message used to exchange addressing
+ * data.
+ */
+int ft_alloc_msgs(struct ct_pingpong *ct)
+{
+	int ret;
+	long alignment = 1;
+
+	/* TODO: support multi-recv tests */
+	if (ct->fi->rx_attr->op_flags == FI_MULTI_RECV)
+		return 0;
+
+	ct->tx_size = ct->opts.options & FT_OPT_SIZE ?
+		  ct->opts.transfer_size : test_size[TEST_CNT - 1].size;
+	if (ct->tx_size > ct->fi->ep_attr->max_msg_size)
+		ct->tx_size = ct->fi->ep_attr->max_msg_size;
+	ct->rx_size = ct->tx_size;
+	ct->buf_size = MAX(ct->tx_size, FT_MAX_CTRL_MSG) + MAX(ct->rx_size, FT_MAX_CTRL_MSG);
+
+	if (ct->opts.options & FT_OPT_ALIGN) {
```

Comment:
Let's just always use aligned data -- remove the non-FT_OPT_ALIGN code path (and remove the CLI option for it, if it still exists).